### DISCUSSION TITLE
compress yolo object detection model (.pt file)

### DISCUSSION THREAD

**Discussion by Aadi4401** [ultralytics/ultralytics#14327] [Q&A]
I've tried compressing my .pt file using pruning, quantization, and various other methods, but these attempts have doubled the file size 20mb file becomes 40mb. If anyone has any ideas on how to effectively compress a .pt file, your help would be greatly appreciated.

**Comment by glenn-jocher** [comment]
@Aadi4401 hello!

Thank you for reaching out. Compressing a YOLO model effectively can indeed be challenging. Here are a few strategies you might find useful:

1. **Quantization**: Ensure you're using post-training quantization correctly. For example, using PyTorch's built-in quantization methods can help:
    ```python
    import torch
    from ultralytics import YOLO

    model = YOLO("path/to/your/model.pt")
    model.fuse()  # Fuse Conv, BN, and ReLU layers
    model.qconfig = torch.quantization.get_default_qconfig('fbgemm')
    torch.quantization.prepare(model, inplace=True)
    torch.quantization.convert(model, inplace=True)
    torch.save(model.state_dict(), "path/to/quantized_model.pt")
    ```

2. **Pruning**: Pruning can reduce the model size, but it needs to be followed by fine-tuning to maintain accuracy. Ensure you're removing the pruned weights properly:
    ```python
    import torch
    from torch.nn.utils import prune

    model = torch.load("path/to/your/model.pt")
    parameters_to_prune = [(module, 'weight') for module in model.modules() if isinstance(module, torch.nn.Conv2d)]
    prune.global_unstructured(parameters_to_prune, pruning_method=prune.L1Unstructured, amount=0.4)
    torch.save(model.state_dict(), "path/to/pruned_model.pt")
    ```

3. **Model Export**: Exporting the model to a different format like ONNX or TensorRT can sometimes result in smaller file sizes and faster inference times:
    ```python
    from ultralytics import YOLO

    model = YOLO("path/to/your/model.pt")
    model.export(format="onnx", simplify=True)
    ```

4. **Check Latest Versions**: Ensure you are using the latest versions of the Ultralytics YOLO package and PyTorch, as updates often include optimizations and bug fixes.

If you continue to face issues, please provide a minimum reproducible example so we can assist you better. You can find guidance on creating one [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Hope this helps! 😊

**Reply by Aadi4401** [reply]
1 Quantization is compressing properly but when i use the output file in object detection i get error 

2 Pruning this method doubles the file size 

3 Model Export also doubles the size 

Quantization worked it decreased the file size but when i use that .pt file for object detection i am getting error



The error

E\.venv\Lib\site-packages\torch\_utils.py:403: UserWarning: TypedStorage is deprecated. It will be removed in the
future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()
  device=storage.device,
Traceback (most recent call last):
  File "C:\Users\Aqua Mick\PycharmProjects\PPE\comp_test.py", line 8, in <module>
    YOLO("quantppe3.pt")
  File "C:\Users\Aqua Mick\PycharmProjects\PPE\.venv\Lib\site-packages\ultralytics\models\yolo\model.py", line 23, in __init__
    super().__init__(model=model, task=task, verbose=verbose)
  File "C:\Users\Aqua Mick\PycharmProjects\PPE\.venv\Lib\site-packages\ultralytics\engine\model.py", line 152, in __init__
    self._load(model, task=task)
  File "C:\Users\Aqua Mick\PycharmProjects\PPE\.venv\Lib\site-packages\ultralytics\engine\model.py", line 241, in _load
    self.model, self.ckpt = attempt_load_one_weight(weights)
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Aqua Mick\PycharmProjects\PPE\.venv\Lib\site-packages\ultralytics\nn\tasks.py", line 813, in attempt_load_one_weight
    model = (ckpt.get("ema") or ckpt["model"]).to(device).float()  # FP32 model
                                ~~~~^^^^^^^^^
KeyError: 'model'

**Reply by glenn-jocher** [reply]
Hello @Aadi4401,

Thank you for providing detailed information about the issues you're facing with quantization, pruning, and model export. Let's address each point:

1. **Quantization Error**: The error you're encountering suggests that the quantized model file might not be correctly structured or saved. To ensure proper quantization and saving, you can follow this example:
    ```python
    import torch
    from ultralytics import YOLO

    model = YOLO("path/to/your/model.pt")
    model.fuse()  # Fuse Conv, BN, and ReLU layers
    model.qconfig = torch.quantization.get_default_qconfig('fbgemm')
    torch.quantization.prepare(model, inplace=True)
    torch.quantization.convert(model, inplace=True)
    torch.save(model.state_dict(), "path/to/quantized_model.pt")
    ```

2. **Pruning Increasing File Size**: Pruning can sometimes increase the file size if not followed by proper fine-tuning and saving. Ensure you are removing the pruned weights correctly:
    ```python
    import torch
    from torch.nn.utils import prune

    model = torch.load("path/to/your/model.pt")
    parameters_to_prune = [(module, 'weight') for module in model.modules() if isinstance(module, torch.nn.Conv2d)]
    prune.global_unstructured(parameters_to_prune, pruning_method=prune.L1Unstructured, amount=0.4)
    torch.save(model.state_dict(), "path/to/pruned_model.pt")
    ```

3. **Model Export Increasing File Size**: Exporting to formats like ONNX or TensorRT should ideally optimize the model for inference. If the file size increases, it might be due to additional metadata or conversion overhead. Ensure you're using the latest versions of the packages and follow the export process correctly:
    ```python
    from ultralytics import YOLO

    model = YOLO("path/to/your/model.pt")
    model.export(format="onnx", simplify=True)
    ```

To help us further diagnose the issue, could you please provide a minimum reproducible example of your code? This will allow us to better understand the context and provide more accurate assistance. You can find guidance on creating one [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Additionally, please verify that you are using the latest versions of the Ultralytics YOLO package and PyTorch, as updates often include important fixes and optimizations.

Looking forward to your response! 😊

**Reply by Aadi4401** [reply]
Hello @glenn-jocher 
you sent the same code again. 
your help would be greatly appreciated.

**Reply by glenn-jocher** [reply]
Hello @Aadi4401,

Thank you for your patience. Let's address the quantization issue you're facing with a more detailed approach to ensure the model is saved correctly and can be loaded without errors.

### Quantization and Saving the Model

Here's a more comprehensive example to quantize and save your YOLO model:

1. **Quantize the Model**:
    ```python
    import torch
    from ultralytics import YOLO

    # Load the model
    model = YOLO("path/to/your/model.pt")
    model.fuse()  # Fuse Conv, BN, and ReLU layers

    # Set the quantization configuration
    model.qconfig = torch.quantization.get_default_qconfig('fbgemm')

    # Prepare the model for quantization
    torch.quantization.prepare(model, inplace=True)

    # Calibrate the model with a few batches of data
    # Assuming you have a DataLoader `data_loader`
    for batch in data_loader:
        model(batch[0])

    # Convert the model to a quantized version
    torch.quantization.convert(model, inplace=True)

    # Save the quantized model
    torch.save(model.state_dict(), "path/to/quantized_model.pt")
    ```

2. **Loading the Quantized Model**:
    ```python
    import torch
    from ultralytics import YOLO

    # Load the quantized model
    model = YOLO("path/to/your/model.pt")
    model.load_state_dict(torch.load("path/to/quantized_model.pt"))

    # Ensure the model is in evaluation mode
    model.eval()

    # Run inference
    results = model.predict("path/to/your/image.jpg")
    ```

### Important Considerations

- **Reproducible Example**: If the issue persists, providing a minimum reproducible example would be highly beneficial. This helps us diagnose the problem more effectively. You can find guidance on creating one [here](https://docs.ultralytics.com/help/minimum_reproducible_example).
- **Latest Versions**: Ensure you are using the latest versions of Ultralytics YOLO and PyTorch. Sometimes, updates include important fixes and optimizations.

If you continue to experience issues, please share more details about your setup and any error messages you encounter. This will help us assist you better.

Thank you for your cooperation! 😊

**Reply by Aadi4401** [reply]
Hello @glenn-jocher 
I do not have a data loader or dataset, only .pt files, which I use for object detection, and they work properly. However, when I compress these .pt files, they do not function correctly. Is there any way to fix this issue so that the compressed files will work properly?
your help would be greatly appreciated.

**Reply by glenn-jocher** [reply]
Hello @Aadi4401,

Thank you for your message. Please ensure you're using the latest versions of Ultralytics YOLO and PyTorch. If the issue persists, try re-saving the quantized model using the following approach:

```python
import torch
from ultralytics import YOLO

# Load and fuse the model
model = YOLO("path/to/your/model.pt")
model.fuse()

# Quantize the model
model.qconfig = torch.quantization.get_default_qconfig('fbgemm')
torch.quantization.prepare(model, inplace=True)
torch.quantization.convert(model, inplace=True)

# Save the quantized model
torch.save(model.state_dict(), "path/to/quantized_model.pt")

# Load the quantized model
model = YOLO("path/to/your/model.pt")
model.load_state_dict(torch.load("path/to/quantized_model.pt"))
model.eval()

# Run inference
results = model.predict("path/to/your/image.jpg")
```

If the problem continues, please check for reproducibility with the latest package versions. Thank you! 😊

**Reply by Aadi4401** [reply]
**upgraded ultralytics and tried this method** 

**got this in terminal** 
While copying the parameter named "model.model.22.cv3.1.2.weight", whose dimensions in the model are torch.Size([10, 256, 1, 1]) and whose dimensions in the checkpoint are torch.Size([10, 256, 1, 1]), an exception occurred : ('Copying from quantized Tensor to non-quantized Tensor is not allowed, please use dequantize to get a float Tensor from a quantized Tensor',).    
        While copying the parameter named "model.model.22.cv3.2.0.conv.weight", whose dimensions in the model are torch.Size([256, 512, 3, 3]) and whose dimensions in the checkpoint are torch.Size([256, 512, 3, 3]), an exception occurred : ('Copying from quantized Tensor to non-quantized Tensor is not allowed, please use dequantize to get a float Tensor from a quantized Tensor',).
        While copying the parameter named "model.model.22.cv3.2.1.conv.weight", whose dimensions in the model are torch.Size([256, 256, 3, 3]) and whose dimensions in the checkpoint are torch.Size([256, 256, 3, 3]), an exception occurred : ('Copying from quantized Tensor to non-quantized Tensor is not allowed, please use dequantize to get a float Tensor from a quantized Tensor',).
        While copying the parameter named "model.model.22.cv3.2.2.weight", whose dimensions in the model are torch.Size([10, 256, 1, 1]) and whose dimensions in the checkpoint are torch.Size([10, 256, 1, 1]), an exception occurred : ('Copying from quantized Tensor to non-quantized Tensor is not allowed, please use dequantize to get a float Tensor from a quantized Tensor',).    
        While copying the parameter named "model.model.22.dfl.conv.weight", whose dimensions in the model are torch.Size([1, 16, 1, 1]) and whose dimensions in the checkpoint are torch.Size([1, 16, 1, 1]), an exception occurred : ('Copying from quantized Tensor to non-quantized Tensor is not allowed, please use dequantize to get a float Tensor from a quantized Tensor',).
        
        
**but it created a compressed file so i tried using that compressed .pt file then i got this error** 

the error 

C:\ppe-detection\myenv\Lib\site-packages\torch\_utils.py:403: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()
  device=storage.device,
Traceback (most recent call last):
  File "C:\ppe-detection\myapp\classes.py", line 3, in <module>
    model1 = YOLO("ppe3new1.pt")
             ^^^^^^^^^^^^^^^^^^^
  File "C:\ppe-detection\myenv\Lib\site-packages\ultralytics\models\yolo\model.py", line 23, in __init__
    super().__init__(model=model, task=task, verbose=verbose)
  File "C:\ppe-detection\myenv\Lib\site-packages\ultralytics\engine\model.py", line 149, in __init__
    self._load(model, task=task)
  File "C:\ppe-detection\myenv\Lib\site-packages\ultralytics\engine\model.py", line 230, in _load
    self.model, self.ckpt = attempt_load_one_weight(weights)
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\ppe-detection\myenv\Lib\site-packages\ultralytics\nn\tasks.py", line 857, in attempt_load_one_weight
    model = (ckpt.get("ema") or ckpt["model"]).to(device).float()  # FP32 model
                                ~~~~^^^^^^^^^
KeyError: 'model'

**Reply by glenn-jocher** [reply]
@Aadi4401 hi there,

Thanks for sharing the details. It looks like the issue stems from loading a quantized model. The error indicates a mismatch between quantized and non-quantized tensors. To resolve this, ensure you're using the latest versions of Ultralytics YOLO and PyTorch. When saving the quantized model, use `torch.save(model.state_dict(), "quantized_model.pt")` and load it with `model.load_state_dict(torch.load("quantized_model.pt"))`. If the problem persists, please try dequantizing the tensors before saving. Let me know if this helps! 😊
